octopoes/octopoes/repositories/origin_repository.py

```diff
@@ -114,9 +114,6 @@ def save(self, origin: Origin, valid_time: datetime) -> None:
         except ObjectNotFoundException:
             old_origin = None
 
-        if old_origin == origin:
-            return
-
         self.session.add((XTDBOperationType.PUT, self.serialize(origin), valid_time))
```

Comment:
Since we are not adding a new origin to the database, Should we not also change the previous origin (or old_origin) to be deleted from this moment in time?

Reply:
Currently an OOI is discovered by a normalizer and added to the database. At that point the origin record is also created, and points to the ooi and the normalizer job that yielded that object.
On subsequent runs on the same boefje/normalizer combination no update would be done to the database, the object would simply keep existing as no valid_till was present. 

To be able to show what that second run 'proofed' we need to make sure we either store the output of each normalizer run somewhere (bytes perhaps), or update the origin record to now point at this most recent proof.
By doing the latter we end up with one of two  things depending on the specific implementation:
* An origin record that when seen trough time/history contains a list of all jobs ever involved in proofing this ooi's existence
* A set of origin records related to this OOI, which all exists end to end, and collectively are the historical record of which jobs are involved in proofing this OOI.
I don't know how origin records are currently implemented, so which one it is needs to be checked, and implemented for accordingly

Reply:
We have some documentation at https://docs.openkat.nl/developer_documentation/octopoes.html#origin
The current Origin model:
https://github.com/minvws/nl-kat-coordination/blob/c51122df60d4556214f898fbdb12072f24fdb6a6/octopoes/octopoes/models/origin.py#L9-L40

Looking at the documentation and the code:
- The Origin model  is used for declarations (objects created through rocky), observations (normalizer output) and inferences (bit output)
- The Origin model has an optional task_id that is only used by normalizers. The `__eq__` of the Origin method currently ignores task_id which causes it to not add a new Origin object if the other fields are the same. The purpose of the Origin objects is to keep track of the origins of objects in the graph and for this we don't need to track every normalizer task, so that hasn't been a problem up to now.
- This proposed change will cause a lot of duplicate inference origins for every bit run that has the same result, so we should look at another way to achiee this.

If we want to track all normalizer tasks in the origins/graph I think we have two options to do that:

1. We change the `__eq__` to include the task_id which will do what the current change intends to do. This will create a new Origin for every normalizer run, but won't cause duplication of inference Origins. It will create a lot more Origins object than we have now and also duplicate observation Origins in the sense that they have the same source/result. I'm not sure whether there isn't code that assumes that that doesn't happen which we would have to fix. Given that it wasn't possible before we also don't have unit tests that would test this, so we should investigate that if implement it this way. There might also be performance issues when we have a lot more Origin objects. 
2. Change the Origin object to include a list of tasks instead of a single task. If the result is the same we just add the task to the lists of tasks of the already existing Origin object. The challenge here is how to change this without breaking existing installations. Given that the task_id currently just points to the first normalizer task that returned this result, I am not sure whether we are really using this information anywhere, so we might be able to just change this without a problems.

Personally I have a preference for option 2, because it fits better with how we use the Origin object and we don't risk performance issues by creatings lots of Origin objects in the database.

Reply:
This has been implemented